aspnetcore/security/authorization/resourcebased/samples/ResourceBasedAuthApp1/Controllers/DocumentController.cs

```diff
@@ -0,0 +1,71 @@
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using ResourceBasedAuthApp1.Data;
+using ResourceBasedAuthApp1.Models;
+using ResourceBasedAuthApp1.Services;
+using System;
+using System.Threading.Tasks;
+
+namespace ResourceBasedAuthApp1.Controllers
+{
+    #region snippet_IAuthServiceDI
+    public class DocumentController : Controller
+    {
+        private readonly IAuthorizationService _authorizationService;
+        private readonly IDocumentRepository _documentRepository;
+
+        public DocumentController(IAuthorizationService authorizationService,
+                                  IDocumentRepository documentRepository)
+        {
+            _authorizationService = authorizationService;
+            _documentRepository = documentRepository;
+        }
+        #endregion
+
+        #region snippet_DocumentViewAction
+        [HttpGet]
+        public async Task<IActionResult> View(Guid documentId)
+        {
+            Document document = _documentRepository.Find(documentId);
+
+            if (document == null)
+            {
+                return new NotFoundResult();
+            }
+
+            if (await _authorizationService.AuthorizeAsync(User, document, Operations.Read))
+            {
+                return View(document);
+            }
+            else
+            {
+                return new ChallengeResult();
+            }
+        }
+        #endregion
+
+        #region snippet_DocumentEditAction
+        [HttpGet]
+        public async Task<IActionResult> Edit(Guid documentId)
+        {
+            Document document = _documentRepository.Find(documentId);
+
+            if (document == null)
+            {
+                return new NotFoundResult();
+            }
+
+            if (await _authorizationService.AuthorizeAsync(User, document, "EditPolicy"))
+            {
+                return View(document);
+            }
+            else
+            {
+                return new ChallengeResult();
```

Comment:
Shouldn't this be Forbid? @HaoK

Reply:
So in 1.x we rely on auto challenge behavior, in 2.0 we should probably use the new IPolicyEvalutor which returns a result which you can use to decide whether to challenge or forbid...

https://github.com/aspnet/Security/blob/rel/2.0.0/src/Microsoft.AspNetCore.Authorization.Policy/IPolicyEvaluator.cs#L38

Reply:
@HaoK Switching to `IPolicyEvaluator` seems to introduce a lot more complexity. What's the downside of using `IAuthorizationService` instead?

Reply:
Well that complexity would still be needed if you want to figure out whether to return Challenge or Forbid.  That's basically what the evaluator gets you.

Reply:
Actually never mind, since this sample is relying on User already being authenticated, the code is fine.